python/cudf/cudf/tests/test_copying.py

```diff
@@ -52,3 +53,309 @@ def test_null_copy():
+    assert_eq(cp_array, cp.array([20, 10, 10, 4, 5]))
+    assert_eq(s2, cudf.Series([20, 10, 10, 4, 5]))
+    assert_eq(s3, cudf.Series([20, 10, 10, 4, 5]))
+
+    s4 = cudf.Series([10, 20, 30, 40, 50])
+    s5 = cudf.Series(s4)
+    assert_eq(s5, cudf.Series([10, 20, 30, 40, 50]))
+    s5[0:2] = 1
+    assert_eq(s5, cudf.Series([1, 1, 30, 40, 50]))
+    assert_eq(s4, cudf.Series([1, 1, 30, 40, 50]))
+
+
+@pytest.mark.parametrize("copy_on_write", [True, False])
+def test_series_str_copy(copy_on_write):
+    cudf.set_option("copy_on_write", copy_on_write)
+    s = cudf.Series(["a", "b", "c", "d", "e"])
+    s1 = s.copy(deep=True)
+    s2 = s.copy(deep=True)
+
+    assert_eq(s, cudf.Series(["a", "b", "c", "d", "e"]))
+    assert_eq(s1, cudf.Series(["a", "b", "c", "d", "e"]))
+    assert_eq(s2, cudf.Series(["a", "b", "c", "d", "e"]))
+
+    s[0:3] = "abc"
+
+    assert_eq(s, cudf.Series(["abc", "abc", "abc", "d", "e"]))
+    assert_eq(s1, cudf.Series(["a", "b", "c", "d", "e"]))
+    assert_eq(s2, cudf.Series(["a", "b", "c", "d", "e"]))
+
+    s2[1:4] = "xyz"
+
+    assert_eq(s, cudf.Series(["abc", "abc", "abc", "d", "e"]))
+    assert_eq(s1, cudf.Series(["a", "b", "c", "d", "e"]))
+    assert_eq(s2, cudf.Series(["a", "xyz", "xyz", "xyz", "e"]))
+    cudf.set_option("copy_on_write", False)
+
+
+@pytest.mark.parametrize("copy_on_write", [True, False])
+def test_series_cat_copy(copy_on_write):
+    cudf.set_option("copy_on_write", copy_on_write)
+    s = cudf.Series([10, 20, 30, 40, 50], dtype="category")
+    s1 = s.copy(deep=True)
+    s2 = s1.copy(deep=True)
+    s3 = s1.copy(deep=True)
+
+    s[0] = 50
+    assert_eq(s, cudf.Series([50, 20, 30, 40, 50], dtype=s.dtype))
+    assert_eq(s1, cudf.Series([10, 20, 30, 40, 50], dtype="category"))
+    assert_eq(s2, cudf.Series([10, 20, 30, 40, 50], dtype="category"))
+    assert_eq(s3, cudf.Series([10, 20, 30, 40, 50], dtype="category"))
+
+    s2[3] = 10
+    s3[2:5] = 20
+    assert_eq(s, cudf.Series([50, 20, 30, 40, 50], dtype=s.dtype))
+    assert_eq(s1, cudf.Series([10, 20, 30, 40, 50], dtype=s.dtype))
+    assert_eq(s2, cudf.Series([10, 20, 30, 10, 50], dtype=s.dtype))
+    assert_eq(s3, cudf.Series([10, 20, 20, 20, 20], dtype=s.dtype))
+    cudf.set_option("copy_on_write", False)
+
+
+def test_dataframe_cow_slice_setitem():
+    cudf.set_option("copy_on_write", True)
+    df = cudf.DataFrame({"a": [10, 11, 12, 13, 14], "b": [20, 30, 40, 50, 60]})
+    slice_df = df[1:4]
+
+    assert_eq(
+        slice_df,
+        cudf.DataFrame(
+            {"a": [11, 12, 13], "b": [30, 40, 50]}, index=[1, 2, 3]
+        ),
+    )
+
+    slice_df["a"][2] = 1111
+
+    assert_eq(
+        slice_df,
+        cudf.DataFrame(
+            {"a": [11, 1111, 13], "b": [30, 40, 50]}, index=[1, 2, 3]
+        ),
+    )
+    assert_eq(
+        df,
+        cudf.DataFrame({"a": [10, 11, 12, 13, 14], "b": [20, 30, 40, 50, 60]}),
+    )
+    cudf.set_option("copy_on_write", False)
```

Comment:
All these restores should have done `cudf.get_option` to figure out what to restore to. (Otherwise imagine the pain when we flip the default).

Reply:
Yup, good call